### ISSUE TITLE
[2.7] possible issue with image width detection for zoom feature

### ISSUE THREAD

**Issue by maximus80** [woocommerce/woocommerce#13290]
In `single-product.js`, in the `init_zoom()` it is checked if the zoom feature can actually be enabled. In order to do so this conditional is used:
```
if ( image_to_zoom.attr( 'width' ) > $( '.woocommerce-product-gallery' ).width() ) {
	zoom_target.trigger( 'zoom.destroy' );
	zoom_target.zoom({
		touch: false
	});
}
```

However the `image_to_zoom` is an objects of matching elements, so the width is actually taken from the first image in the object. That will usually be the product image. So this means that if he product image is not large enough for zoom to work, then the feature will be completely disabled. Even if all other images set in product gallery setting would be large enough.

This becomes more problematic if `animationLoop` is used in the gallery flexslider. That is disabled by default, but some themes will use it. If that is used, then the first image in the object of matched elements is no longer the product image, but the last image set in product gallery setting. Which more or less turns the whole check upside down.

Would it be possible to do the check on a per image basis and enable the zoom for every image that is large enough and just have it disabled on the ones that are too small?

**Comment by mikejolley** [comment]
https://github.com/woocommerce/woocommerce/issues/13278

> Zoom has to be init once, not init per image. So we only have the choice of on/off. It makes more sense to base this on the main product image than supporting gallery images IMO.

**Comment by maximus80** [comment]
Hm yeah it there is no chance of doing it on per image basis, which would really be a huge advantage, then it does make most sense to base on main product image.

Just thought I bring it up, as the behaviour gets odd, when `animationLoop` param is set to true for flexslider, through the `woocommerce_single_product_carousel_options` filter.

**Comment by mikejolley** [comment]
What does that do exactly in this context?

**Comment by maximus80** [comment]
As said above, if `animationLoop` is set t true, the first image in the object of matched elements for the width check changes.
Reason is, that this parameter allows the flexslider to loop. So also on the first displayed image, the "back" navigation button will be there. In order to allow a smoother transition. flexslider adds the last image of the wrapping container before the first displayed image.

So in our case, the product image no longer is the first image, but the last image that was added to the gallery images is. This in turns changes the object of matched elements, where now also that image is the first, and thus its width is used to determine if zoom can be used or not.

In a nutshell: instead of the main product image, the last image in the product gallery setting will decide if zoom can be used or not.

**Comment by mikejolley** [comment]
Ok, so using consistent images would be ideal in cases like that. Not sure zoom can account for it.

**Comment by maximus80** [comment]
Yes that would be advisable in most use cases really.

I understand this is problematic to fix, if the zoom can only be bulk activated.

It just does not seem straight forward to the user, what is preventing the zoom feature to work. Especially when most pictures have correct sizes. Makes it kind of bad situation.

**Comment by lukeabeck** [comment]
Yes using consistent size images is ideal, and what a good design would yield. However a lot of people use varying image sizes.

From usability standpoint, this will confuse the % of users who do not use all equal sizes and create a lot more support tickets and questions in general.

If zoom can account for that in someway, that would be ideal.

**Comment by mikejolley** [comment]
@jameskoster what do you think?

**Comment by jameskoster** [comment]
It's a shame we can't do it per image as that would be ideal.

For the animationLoop issue couldn't we just be more specific with the selector we use to find the image to check against rather than returning an object? So;

`var zoom_target   = $( '.woocommerce-product-gallery__image' );`

becomes

`var zoom_target   = $( '.woocommerce-product-gallery__image.flex-active-slide' );`

On load the active slide is always the main featured image.

**Comment by maximus80** [comment]
This would make most sense here, I think. The additional specificity will fix the animationLopp issue and will also simplify the jQuery check, as just the correct element would be selected instead of all images. So that would already help.

If it were possible to do it per image in the future that would be even better then.

**Comment by lukeabeck** [comment]
Thanks guys 👍
